### DISCUSSION TITLE
hardware back button support for bottoms

### DISCUSSION THREAD

**Discussion by aliraza-noon** [wix/react-native-navigation#7365] [Enhancements & New Ideas]
## 🐛 Bug Report

When switching between bottom tabs it does not go back to previously selected tab when i press hardware back button, please check popular apps like

* youtube
* facebook
* twitter

### Have you read the [Contributing Guidelines on issues](https://wix.github.io/react-native-navigation/docs/meta-contributing/)?

Yes

## To Reproduce

(Write your steps here:)

1. Setup a demo app with 2 bottoms 
2. Tap on 1st tab first and then switch to 2nd tab
4. Try pressing hardware back button

## Expected behavior

It should take you back to previously selected tab

## Actual Behavior

It closes the app when you press the hardware back button

## Your Environment

- React Native Navigation version: 7.23.1
- React Native version: 0.66.0
- Platform(s) (iOS, Android, or both?): Android
- Device info (Simulator/Device? OS version? Debug/Release?): Pixel2 api version 30 debug

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✖️  &nbsp; Yes, I have the time, and I know how to start.
 - ✖️  &nbsp; Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✅  &nbsp; No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by phuongwd** [comment]
I have handled that on my app. Im prefer the behavior "always to the first (default) tab" when press on hardware back button.
Just checked, the Material Design no longer mentions back button behavior.

| Popular apps  | back behavior |
| ------------- | ------------- |
| Google Photos | quit app on back press on any tab  |
| Youtube, Facebook | previously selected tab |
| Airbnb, Google+, Play Books | pressing back takes you always to the first (default) tab, pressing again closes the app |

**Comment by swabbass** [comment]
This should be converted to a discussion (on the side menu "convert to discussion" as an idea, and maybe you can suggest adding a new option for hardware back like:
```javascript
const options = {
  hardwareBackButton:{
    dismissModalOnPress: true, //exists
    popStackOnPress: true, // exists
    bottomTabsBehaviour: 'exit' | 'previous' | 'first', // new
  }
}
```

**Reply by aliraza-noon** [reply]
@swabbass to keep it inline with previous properties it should be called `bottomTabsOnPress`, what do you think ?

**Reply by swabbass** [reply]
Yeah it could be, but how about if we can achieve your described behaviour in another way:
What we can infer from docs, that we can achieve such behaviour, [see](https://wix.github.io/react-native-navigation/docs/customBackNavigation#handling-the-hardware-back-button), also [see](https://wix.github.io/react-native-navigation/docs/bottomTabs/#selecting-a-tab-by-index).
1. Set Options to be only for the bottom tabs screen (at bottom tabs level):
```javascript
options: {
  hardwareBackButton: {
    dismissModalOnPress: false,
    popStackOnPress: false,
  }
}
```
2. set navigation button callback in the bottom tabs screen where you want such behaviour as follows:
```javascript
navigationButtonPressed({ buttonId }) {
  if (buttonId === 'RNN.hardwareBackButton') {
    console.log('The hardware back button was pressed!'); 
    Navigation.mergeOptions(this.props.componentId, {
          bottomTabs: {
    currentTabIndex: 0
  }
});
  }
}
```
does this help?

**Reply by aliraza-noon** [reply]
well we are able to do this, the hard part comes when you want to switch back to previously selected tab then you have to keep track of the last selected tab and update it every time tabs change and then restore when the hardware back button is tapped it becomes ugly very quickly when it involves bit more logic like deep links and navigation

**Reply by swabbass** [reply]
got it, so, it seems to me that we can do it  natively by adding `bottomTabsOnPress: 'exit' | 'previous' | 'first'`, in which, while on a bottom tabbed screen, no modals, and no stack to pop, hardware back button will do one of the follows:

- Exit the app (default behaviour as today).
- Whatever tab is selected, select `currentTabIndex-1` until 0, and then the next back will do exit.
- Whatever tab is selected, select the first tab, and the next back will do exit.

A starting point for this is modifying `BottomTabsController.java#handleBack`, if you have the time to contribute then it would be awesome, I will give a hand on this, if not I'll schedule it, and it could be implemented soon.

**Reply by aliraza-noon** [reply]
> Exit the app (default behaviour as today).
> Whatever tab is selected, select currentTabIndex-1 until 0, and then the next back will do exit.
> Whatever tab is selected, select the first tab, and the next back will do exit.

one additional flow would be suppose I have 4 tabs 
1. Home
2. Search
3. Categories
4. User Profile

If i switch to **User Profile** from **Home** when hardware back button is pressed I would like to return to last selected tab which in this case would be **Home**.

> A starting point for this is modifying BottomTabsController.java#handleBack, if you have the time to contribute then it would be awesome, I will give a hand on this, if not I'll schedule it, and it could be implemented soon.

I would love to contribute but unfortunately we are near a beta launch don't have the time right now. if you schedule it if time permits I can work on it if it is not picked up by someone else

**Reply by swabbass** [reply]
I'll schedule it for me to work on it soon.

**Reply by swabbass** [reply]
@aliraza-noon : https://github.com/wix/react-native-navigation/pull/7380

**Comment by aliraza-noon** [comment]
@swabbass I don't see the "convert to discussion"
